src/main/java/core/Parser.java

```diff
@@ -0,0 +1,212 @@
+            return new String[]{fromSplit[0].trim(), toSplit[0].trim(), toSplit[1].trim()};
+        }
+
+        case DEADLINE: {
+            // Check for /by keyword
+            String[] bySplit = this.splitByDelimiter(this.userInput, "/by", 3);
+
+            // Check for descriptor
+            if (bySplit[0].isEmpty()) {
+                throw new exceptions.missing.Parameter(this.keyword);
+            }
+
+            return new String[]{bySplit[0].trim(), bySplit[1].trim()};
+        }
+
+        default:
+            throw new exceptions.invalid.Input("DIO - extractTaskParams - used unexpectedly");
+        }
+
+    }
+
+    /**
+     * Splits String based on delimiter.
+     * Example:
+     * splitByDelimiter("Gymkhana /from TIME /to TIME","/from",2)
+     * returns ["Gymkhana","TIME /to TIME"]
+     *
+     * @param in        The string to split
+     * @param delimiter The delimiter to watch for
+     * @param limit     What is the maximum times the method should split
+     * @return String array with limit number of entries
+     * @throws DukeException Thrown when missing parameters or invalid input is detected
+     */
+    private String[] splitByDelimiter(String in, String delimiter, int limit) throws DukeException {
+        String[] ret = in.split(delimiter, limit);
+        if (ret.length < 2) {
+            throw new exceptions.missing.Parameter(this.keyword);
+        } else if (ret.length > 2) {
+            throw new exceptions.invalid.Input(
+                    String.format("Multiple `%s` detected, only one is allowed, please try again.", delimiter));
+        }
+        return ret;
+    }
+
+    /**
+     * Parses save file and loads it into the given TaskMaster
+     *
+     * @param task The task stored in the safe file
+     * @param tm   The runtime TaskMaster object
+     * @throws DukeException Thrown when input is invalid, which can appear when user manually edits the file.
+     */
+    public static void parseSaveFile(String[] task, TaskMaster tm) throws DukeException {
+        switch (task[0]) {
+        case "T":
+            tm.addToDo(task[1], Boolean.parseBoolean(task[2]));
+            break;
+        case "D":
+            tm.addDeadLine(task[1], Boolean.parseBoolean(task[2]),
+                    DateHandler.convert(task[3]));
+            break;
+        case "E":
+            tm.addEvent(task[1], Boolean.parseBoolean(task[2]),
+                    DateHandler.convert(task[3]), DateHandler.convert(task[4]));
+            break;
+        default:
+            throw new exceptions.invalid.Input(String.format("Unknown command for %s", task[0]));
+        }
+    }
+
+    /**
+     * Parse user input into meaningful commands.
+     * @param tm The runtime TaskMaster object
+     * @return Message from the corresponding commands to print out.
+     * @throws DukeException Thrown when missing parameters is found or index error etc.
+     */
+    public String parse(TaskMaster tm) throws DukeException {
```

Comment:
Really neat and easy to read 👍, also followed convention and carefully avoided indentation for case clauses.

Reply:
Thank you for your kind words!